File: de.unistuttgart.iste.sqa.mpw.hamstersimulator.cpp/core/test/view/ViewModelStringifier.cpp
#include "ViewModelStringifier.h"

#include "CollectionHelpers.hpp"

#include <algorithm>

namespace hamsterviewmodel {

ViewModelStringifier::ViewModelStringifier(
        const std::unordered_map<std::string, std::string>& imageCharsMapping, int maxCharsPerCell)
    : imageCharsMapping(imageCharsMapping)
    , maxCharsPerCell(maxCharsPerCell)
{
}

std::string ViewModelStringifier::territoryToExpectationString(const GameViewModel& viewModel) {
    auto& nonConstViewModel = const_cast<GameViewModel&>(viewModel); // TODO const-correctness: adapt after migrate generator
    std::string actual;
    int height = nonConstViewModel.getHeight();
    for (int y = 0; y < height; y++) {
        std::string rowString = rowToString(viewModel, y);
        actual.append(rowString);
        actual.append("\n");
    }
    return actual;
}

std::string ViewModelStringifier::rowToString(const GameViewModel& viewModel, int y) const {
    auto& nonConstViewModel = const_cast<GameViewModel&>(viewModel); // TODO const-correctness: adapt after migrate generator
    std::string result = "|";
    int width = nonConstViewModel.getWidth();
    for (int x = 0; x < width; x++) {
        auto cell = nonConstViewModel.getCellAt(y, x);
        result.append(cellToString(*cell));
        result.append("|");
    }
    return result;
}

std::string ViewModelStringifier::cellToString(const ViewModelCell& cell) const {
    auto& nonConstCell = const_cast<ViewModelCell&>(cell); // TODO const-correctness: adapt after migrate generator
    std::string result;
    for (const auto& layer : nonConstCell.getLayers()) {
        auto& nonConstLayer = const_cast<ViewModelCellLayer&>(*layer);
        if (nonConstLayer.getVisible()) {
            std::string key = nonConstLayer.getImageName();
            int rotation = nonConstLayer.getRotation();
            if (rotation != 0) {
                key.append("[");
                key.append(std::to_string(rotation));
                key.append("]");
            }
            result.append(getImageCharMapping(key));
        }
    }
    int length = static_cast<int>(result.length());
    int spaces = std::max(0, maxCharsPerCell - length);
    for (size_t i = 0; i < spaces; i++) {
        result.append(" ");
    }
    return result;
}

std::string ViewModelStringifier::logToString(const GameViewModel& viewModel) {
    auto& nonConstViewModel = const_cast<GameViewModel&>(viewModel); // TODO const-correctness: adapt after migrate generator
    std::string actual;
    for (auto& entry : nonConstViewModel.getLogEntries()) {
        actual.append(entry);
        actual.append("\n");
    }
    return actual;
}

const std::string& ViewModelStringifier::getImageCharMapping(const std::string& key) const {
    auto iter = imageCharsMapping.find(key);
    return (*iter).second;
}

}

File: de.unistuttgart.iste.sqa.mpw.karasimulator.java/core/src/test/java/de/unistuttgart/kara/view/ViewModelStringifier.java
package de.unistuttgart.kara.view;

import de.unistuttgart.kara.viewmodel.GameViewModel;
import de.unistuttgart.kara.viewmodel.ViewModelCell;
import de.unistuttgart.kara.viewmodel.ViewModelLogEntry;

import java.util.Map;

public class ViewModelStringifier {
    private final Map<String, String> imageCharsMapping;
    private final int maxCharsPerCell;

    public ViewModelStringifier(Map<String, String> imageCharsMapping, int maxCharsPerCell) {
        this.imageCharsMapping = imageCharsMapping;
        this.maxCharsPerCell = maxCharsPerCell;
    }

    public String worldToExpectationString(final GameViewModel viewModel) {
        final StringBuilder actual = new StringBuilder();
        final var height = viewModel.getSize().getRowCount();
        for (int rowIndex = 0; rowIndex < height; rowIndex++) {
            var s = rowToString(viewModel, rowIndex);
            actual.append(s);
            actual.append("\n");
        }
        return actual.toString();
    }

    private String rowToString(final GameViewModel viewModel, final int y) {
        final StringBuilder s = new StringBuilder("|");
        final var width = viewModel.getSize().getColumnCount();
        for (int columnIndex = 0; columnIndex < width; columnIndex++) {
            var cell = viewModel.getCellAt(y, columnIndex);
            s.append(cellToString(cell));
            s.append("|");
        }
        return s.toString();
    }

    private String cellToString(final ViewModelCell cell) {
        StringBuilder cellStringBuilder = new StringBuilder();
        for (var layer : cell.getLayers()) {
            if (layer.getVisible()) {
                var key = layer.getImageName();
                var rotation = layer.getRotation();
                if (rotation != 0) {
                    key += "[" + rotation + "]";
                }
                cellStringBuilder.append(imageCharsMapping.get(key));
            }
        }
        final var length = cellStringBuilder.length();
        final var spaces = maxCharsPerCell - length;
        cellStringBuilder.append(" ".repeat(Math.max(0, spaces)));
        return cellStringBuilder.toString();
    }

    public static String logToString(final GameViewModel viewModel) {
        StringBuilder actual = new StringBuilder();
        for (ViewModelLogEntry entry : viewModel.getLogEntries()) {
            actual.append(entry.getMessage());
            actual.append("\n");
        }
        return actual.toString();
    }

    public static String buttonBarToExpectationString(final GameViewModel viewModel) {
        final StringBuilder actual = new StringBuilder();
        actual.append(buttonToString("play", viewModel.getPlayButtonEnabled()));
        actual.append(" ");
        actual.append(buttonToString("pause", viewModel.getPauseButtonEnabled()));
        actual.append(" ");
        actual.append(buttonToString("undo", viewModel.getUndoButtonEnabled()));
        actual.append(" ");
        actual.append(buttonToString("redo", viewModel.getRedoButtonEnabled()));
        return actual.toString();
    }

    private static String buttonToString(final String name, final boolean enabled) {
        if (enabled) {
            return "[" + name + "]";
        }
        return "/" + name + "/";
    }

}
